packages/financial-templates-lib/src/helpers/acrossFeesCalculator.ts

```diff
@@ -0,0 +1,89 @@
+// This Util calculates the across realized LP fees. See https://gist.github.com/chrismaree/a713725e4fe96c531c42ed7b629d4a85
+// gist for a python implementation of the logic in this file.
+// TODO: add a link to chase's notebooks once they are production ready to explain how the logic works past the comments included here.
+
+import BN from "bn.js";
+import Decimal from "decimal.js";
+import Web3 from "web3";
+
+const { toBN, toWei, fromWei } = Web3.utils;
+const toBNWei = (number: string | number) => toBN(toWei(number.toString()).toString());
+const fixedPointAdjustment = toBNWei("1");
+
+export interface RateModel {
```

Comment:
Can you clarify what this "rate model" is? It appears to me that its a function that outputs interest rate and takes utilization % as input? I didn't realize it could be visualized as a graph until reading later code.

Reply:
ye, in the future we'll hopefully have a medium article we can link to explaining more. the visual representation is this:

![image](https://user-images.githubusercontent.com/12886084/134341864-5532c9bb-362f-4a60-a22d-9e34f4ef90bf.png)

I'll work on adding better/more comments.

Reply:
Isn't there a tool you can use to "draw" an image in comments to put into code?

Reply:
Nevermind graphs are hard to transcribe to ascii :):

Reply:
```                                                                                                                                                                                                                                                                             
                                                    oX.                                                                                                                                                                                                                                                                                 
                                                    oX.                                                                                                                                                                                                                                                                                 
                                                    oX.                                                                                                                                                                                                                                                                                 
                                                    oX.                                                                                                                                                                                                                                                                                 
                                                    oX.                                                                                                                                                                                                                                                                                 
                                                    oX.                                                                                                                                                                                                                                                                                 
                                                    oX.                                                                                           ."""_.                  '"""_                   :"";,                                                                                                                                 
                                                    oX.                                                                                           .dn"]F.         "'      ,Z["5n         ."       ~%|"Y[        ;v"J..  ..    .. .. /|  .     .. J;`  ..  .. ..  .`   ./[                                                               
                                                    oX.                                                                                           _Z|vL"`;;.   `"~Ji~:    iP"fu_`!.    :"|I""'    Jk"oe-_~;     cZr%vTtuLL2  /P?\5p[ud?YuC/ .u}}|Uyyu_56orgio}fT|YuqtncLL                                                               
                                                    oX.                                                                                          _Ln.rP~n!ft      r_     ;$[ oT" o_      ,t     .\S! Ya;;\j _~  ~CiL e}nn5}  /L u6LLvTcqon; .5[f|L;"L[L5C[y.u[x$iYyn|Tl55                                                               
                                                    oX.                                                                                          ...  :`/\t:      :.     ... .:.`f~      .:      ... .:.i[\ -;   ... .  --    .  ,.. . .:.    '. .  . :.... |zxj .:.  :..                                                               
                                   \ljJ_tl|         oX.                                                                                                                                                          .         . . .           .                                                                      \Li-                  
                                    `oclL.8~  !TTTTTmWwTTTTe                                                                                                         `\|-in-  ~L[ ro[a;ulo\xf\!!    "!\..!i: /i__n  .; _;2vv/L\j:||;`\|-|L~[.~i;."!\.  /iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiEMWSj`               
                                    u] i6"p;        oX.                                                                                                              }LdTvF`   Io 8~ FLn Ccslxe?t. -Pi[LLx!pJa/CvT  /w u]C[nvht6:jS![LbwiS!6au~6sPi5o                                                             \SLLo/.               
                                    /.  /v;         oX.                                                                                                              Luy6|b:   lf [u}o"Lj5` |\rIu; -Pjyxfolniusn!5  _Tshtyfyin|n|Pj/nuJd/6|n?5vn!2.ia                                                             `"x|.                 
                                                    oX.                                                                                                               . .          ..   .   .      .x..  ..  ..       .              . .     ..                                                                  _j['                   
                                                    oX.                                                                                                                                                                                                                                                        -rf~                     
                                                    oX.                                                                                                                                                                                                                                                      .|x\.                      
                                                    oX.                                                                                                                                                                                                                                                    `ie!.                        
                                                    oX.                                                                                                                                                                                                                                                  .\x|.                          
                                                    oX.                                                                                                                                                                                                                                                 ~fr-                            
                                                    oX.                                                                                                                                                                                                                                               _[j_                              
                                                    oX.                                                                                                                                                                                                                                             :s}"                                
                                                    oX.                                                                                                                                                                                                                                           `ix!.                                 
                                   :jsr.rl|         oX.                                                                                                                                                                                                                                         .\x|.                                   
                                   L6j5fL.8/  !TTTTTmWTTTTTe                                                                                                                                                                                                                                   "f?-                                     
                                   Lu"SJ6"p;        oX.                                                                                                                                                                                                                                      'rj_                                       
                                    ;/` ,~.         oX.                                                                                                                                                                                                                                    ,v}"                                         
                                                    oX.                                                                                                                                                                                                                                  .ix!.                                          
                                                    oX.                                                                                                                                                                                                                                .!xi.                                            
                                                    oX.                                                                                                                                                                                                                               "f?-                                              
                                                    oX.                                                                                                                                                                                                                             _[j_                                                
                                                    oX.                                                                                                                                                                                                                           ,v}"                                                  
                                                    oX.                                                                                                                                                                                                                         .ie!.                                                   
                                                    oX.                                                                                                                                                                                                                       .!xi.                                                     
                                                    oX.                                                                                                                                                                                                                      "f?-                                                       
                                                    oX.                                                                                                                                                                                                                    /et`                                                         
                                   ~ov\'fc[.        oX.                                                                                                                                                                                                                  ,vl:                                                           
                                   [njxIn F!  !TTTTTmWkTTTTx                                                                                                                                                                                                           .ix!                                                             
                                   t[rq?6?k:        oX.                                                                                                                                                                                                              .!xi.                                                              
                                   .;". -~.         oX.                                                                                                                                                                                                             "}v,                                                                
                                                    oX.                                                                                                                                                                                                           _j[_                                                                  
                                                    oX.                                                                                                                                                                                                         -?}"                                                                    
                                                    oX.                                                                                                                                                                                                       .ix!.                                                                     
                                                    oX.                                                                                                                                                                                                     .!xi.                                                                       
                                                    oX.                                                                                                                                                                                                    "}s,                                                                         
                                                    oX.                                                                                                                                                                                                  ~fr'                                                                           
                                                    oX.                                                                                                                                                                                                ;[j:                                                                             
                                                    oX.                                                                                                                                                                                              .lz;                                                                               
                                                    oX.                                                                                                                                                                                            ./x/.                                                                                
                                    :nt`xc[.        oX.                                                                                                                                                                                           ;}s:                                                                                  
                                   |uTy[n F\  !TTTTTmKTTTTTf                                                                                                                                                                                    _j[;                                                                                    
                                   irkki6lw:        oX.                                                                                                                                                                                       -rf~.                                                                                     
              .zo, |2oa"[l ?[.       .. -".         oX.                                                                                                                                                                                     .|x\.                                                                                       
              [OEL 5N[N2~mPm/                       oX.                                                                                                                                                                                   ./ei`                                                                                         
             |@Tn@x5X!;  !Q|                        oX.                                                                                                                                                                                 .\x|.                                                                                           
             _"  ;"_"    .!.                        oX.                                                                                                                                                                                ~e?-                                                                                             
                                                    oX.                                                                                                                                                                              _[j_                                                                                               
                                                    oX.                                                                                                                                                                            :s}"                                                                                                 
                                                    oX.                                                                                                                                                                          .ix!.                                                                                                  
                                                    oX.                                                                                                                                                                         ;]|.                                                                                                    
                                                    oX.                                                                                                                                                                       :j[;                                                                                                      
                                                    oX.                                                                                                                                                                     `re~.                                                                                                       
                                   iI[[,Ijz.   .....uX......                                                                                                                                                              .|x|.                                                                                                         
                                   :_I8o2 C\  /LLLLLEKLLLLL]                                                                                                                                                            .ix!.                                                                                                           
                                   }Iln\Lju.        oX.                                                                                                                                                               .!xi.                                                                                                             
                                    ..   .          oX.                                                                                                                                                              "f?,                                                                                                               
                                                    oX.                                                                                                                                                            _[j_                                                                                                                 
                                                    oX.                                                                                                   .~/////".                                              ,v}"                                                                                                                   
                                                    oX.                                                  .;;;;::         .,,`.  '.    .`  -,'             .i5|:|x"                                             .ix!.                                                                                                                    
                                                    oX.                                                  .io%r!gvv_`ir~  "u2oc"/j|!is;zI;!uny    :r?"-ti.  zL  i\                                            .!xi.                                                                                                                      
                                                    oX.                                                    !p .U~[LTkzp_  ``uZS_5[8iik}ZP"```    di:8jP_  .8/ .l                                            :}?-                                                                                                                        
                                                    oX.                                                    "o .J./}rall.    lis[[|I:;Ii?ie.  \t  jxjx/f   .}i~\:                                          `te_                                                                                                                          
                                                    oX.                                                                                  ;\~;":                                                         -?f"                                                                                                                            
                                                    oX.                                                                                  :~!|iiiiiiittiiti||!"_-.                                     .ix!.                                                                                                                             
                                     .              oX.                                                                                                 .`:"/\iiiiiiiitiiiiii\/;:`.          .r"    .!xi`                                                                                                                               
                                   ?ocL!nc5`  :iiiiiY@tiiii~                                                                                                              .-_"!|iiiitiiiiiitiibQEu/"}s,                                                                                                                                 
                                   _r}ay2 C\  "JJJJJZDaJJJIr                                                                                                                               ';t%QNS2n;                                                                                                                                   
                                   Jdjr/nju.        oX.                                                                                                                             .:;/\|||!|!,.  :                                                                                                                                    
                                         .          oX.                                                                                                                     .`_"!||||!"_`          }.                                                                                                                                   
                                                    oX.                                                                                                              .,_~\||||/;:.                 _                                                                                                                                    
                                                    oX.                                                                                                       .,_~\|||\~_,.                        ].                                                                                                                                   
                                                    oX.                                                                                                .:;/\|||\~;,.                               _                                                                                                                                    
                                                    oX.                                                                                        `_~!||||!":.                                        t.                                   .\v!\"                  ./v!!~                                                                  
                                                    oX.                                                                                .-_"!||||~;:.                                               /.                                    yL "E;         \:       [6 _%/        |:s~                |.          "l                     c_
                                                    oX.                                                                         .:;~\|||\~;,.                                                      i.                                   -Z?tb".~/:   .~~yt~"    .Sj|b!."!     ,Z[LLzu~nfL!  o5/JIL/ZtIooI  ~nlxjZvL"ff6;Lav5sTtayoeius%/
                                                    oX.                                                                  .:"!||||!"_-.                                                             !.                                  ,uS`/m~Iz\o    ..l"..   -xZ,_m\_"o  "   TtLt`6[%xYi  T|;6oX~Y|P55n. lT"cjL n[paOr6;cT"$eZu555o"P/
                                                    oX.                                                          .`_"!||||!":.                                                                     i.                                  :""..~"!jtt      "`     -""' "~.!o` z   ;/", ".~r!   "-.i|\;"~:ii,   /t"," ;:|i!""._LTS/:ti-.|i\`
                                                    oX.                                                   .-_~!||||/":.                                                                            \.   _:                                                                                                                 ,;`          
                                    "ny r}y}  _tttttC@?tttt!                                       .-_"\|||\~;,.                                                                                   i;sLmZ;.......................................                                                     _j[r!jj~                          
                                     ln d.;F. ;xxxxxPDJxxxxt                                .:;/\|||\~;'.                                                                                          rfU@MZiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii!.               .:..j_  |iI~:  .:.   f:..|'.`'`x ..  .bL:.av.                          
                                     iI t}yj        oX.                              `_"!||||!"_-.                                                                                                 |. .!f"                                                      _nnS!Zi .LwduJniFn8/  6L5;L|SsT\S5I,  :P_ `s                            
                                                    oX.                     .'_"!|||\~;:.                                                                                                          v.                                                           vp]S|Li  lL5~_ntwxu_  nin|n|n siL\L;  "L//|-                            
                                                    oX.              .,;~\|||\/;:.                                                                                                                 "                                                             .... .   .     .`                     ._.                              
 ~v/!"                                              oX.       .:"!||||!"_-.                                                                                                                        l.                                                                                                                                   
 jL !Z:                                        |Lj; oX..:"/\|||!"_`                                                                                                                                ;                                                                                                                                    
.$?iq:_\\-          .iiiiiiiiiiiiiiiiiiiiiiiiiirmMQ%ZSi|!"_`.                                                                                                                                      j.                                                                                                                                   
[d:"S|5izr                                     |Unr:oX.                                                                                                                                            ;                                                                                                                                    
    ..;|i:                                     .    oX.                                                                                                                                            [.                                                                                                                                   
                                                    oX.                       ,;                        l!                       \l                        }n                         ;'           _           ;x`                      .z"                       :Ui                     oo                        S|                  
                                                    oX.                       uX.                      .Eu                       nE                        LZ                        .NI           [.          sW;                      :@z                       _R[                     pY                        Kl                  
                                                    oX.                       uX.                      .Eu                       nE                        LZ                        `NI           '           sW;                      :@z                       _R[                     pY                        Kl                  
                                        /jjjjjjjjjjjU@[jjjjjjjjjjjjjjjjjjjjjjjS@[jjjjjjjjjjjjjjjjjjjjjj[@SjjjjjjjjjjjjjjjjjjjjjjjP@jjjjjjjjjjjjjjjjjjjjjjjjZNjjjjjjjjjjjjjjjjjjjjjjjj}RFjjjjjjjjjjjg[jjjjjjjjjjTQojjjjjjjjjjjjjjjjjjjjjjIWYjjjjjjjjjjjjjjjjjjjjjjjaQ6jjjjjjjjjjjjjjjjjjjjjmEjjjjjjjjjjjjjjjjjjjjjjjjQTjjjjjjjjjjjjjjs`  
                                        /ccccccccccjU@[ccccccccccccccccccccccjS@[cccccccccccccccccccccc[@ScccccccccccccccccccccccP@jcccccccccccccccccccccccZNjccccccccccccccccccccccc}RpccccccccccccjccccccccccTQoccccccccccccccccccccccxWYjccccccccccccccccccccccJQ6cjccccccccccccccccccjmEjccccccccccccccccccccccjQTccccccccccccccv`  
                                                    oX.                       uX.                      .Eu                       nE                        LZ                        .NI                       sW;                      :@z                       _R[                     pY                        Kl                  
                                                    oX.                       uX.                      .SJ                       oS                        vJ                        `NI                       tN;                      ,mc                       -L!                     ][                        L/                  
                                                    oX.                       ;|                        ..                       ..                       __ :!_                     .|;                        .                        .                      -!". "/.               .//. :!_                 `:  /". ;!,             
                                                    oX.                    ."[` tl?.                 |uxy\ulu,                ro[u_ulu!                  ?Zo/Y|Ll                  ijv"_[[~                  ulJ"x]al                `?lTxrojo.                .FyU[uorh,              Leld?b|b,               "Tn JsILzL|q"            
                                                    oX.                    `rZ_\Y.6t                 _iJ]Tv wi                :_}SIL jn                 ILkYIL fu                  Tf[!Tr~C                 _Zs5zC."F                  [o.Y/ F~                ;U[LuY/.C;              [efZ5o u;                ln k."CLl Ti            
                                                    |a                      .F_"p~q~                 [8s\tnjn:                [asL\Lcn\                  .\i`]je:                 .z~}b5yjL                 .Lsniuzo[                 "L. ju[u.                `}cf\!x[r               i[[|,fj}                 /r "zf!;xjj.            
                                                                             -  .".                                                                                                ."_  __                                                                                                                                              
                                                                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                                        
                                                                                                                                                                          .\i|;          ;;  .\` ~;.- !:"_;"        .: ";                                                                                                               
                                                                                                                                                                          "WLoRviout.joo!L6  /Q| ZTTN\Ui82IL\Jui|ooieRrLI;yoI:tJoI                                                                                                      
                                                                                                                                                                          "WYu5?&rtKL@"aNL6  ~Wt.%na%,Oc82n8;ZY~d6Z%i@;8n$h:%nCh/R.                                                                                                     
                                                                                                                                                                          :T"  .Jnna_52Ljl[  .xLnL"!T\5!e?ve[TJt5uuu;Tixv\LuL"}s'L.                                                                                                     
                                                                                                                                           
```

Reply:
@nicholaspai how did you generate that? lol

Reply:
[here](https://cloudapps.herokuapp.com/imagetoascii/) lol